File: go-version/leetcode/LC3306/count_of_substrings.go
package LC3306

import "strings"

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/3/13 下午11:30
 * @FilePath: count_of_substrings
 * @Description: 可以将恰好型问题转换成2个至少型问题
 */

func countOfSubstrings(word string, k int) int64 {
	return calculate(word, k) - calculate(word, k+1)
}

func calculate(ss string, k int) int64 {
	var ans int64
	vowelCntMap := make(map[byte]int)
	consonantCnt := 0
	low := 0
	for i := 0; i < len(ss); i++ {
		if strings.ContainsRune("aeiou", rune(ss[i])) {
			vowelCntMap[ss[i]]++
		} else {
			consonantCnt++
		}

		leftChar := ss[low]
		for len(vowelCntMap) >= 5 && consonantCnt >= k {
			if vowelCntMap[leftChar] > 0 {
				vowelCntMap[leftChar]--
				if vowelCntMap[leftChar] <= 0 {
					delete(vowelCntMap, leftChar)
				}
			} else {
				consonantCnt--
			}
			low++
		}
		ans += int64(low)
	}
	return ans
}

File: python-version/LC3306/count_of_substrings.py
class Solution:
    def countOfSubstrings(self, word: str, k: int) -> int:
        return self.__calculate(word, k) - self.__calculate(word, k+1)

    @staticmethod
    def __calculate(ss: str, k: int) -> int:
        vowel_cnt_map = {}
        consonant_cnt = 0
        ans, low = 0, 0
        for ch in ss:
            if ch in "aeiou":
                vowel_cnt_map[ch] = vowel_cnt_map.get(ch, 0) + 1
            else:
                consonant_cnt += 1

            left_char = ss[low]
            while len(vowel_cnt_map) >= 5 and consonant_cnt >= k:
                if vowel_cnt_map.get(left_char, 0) > 0:
                    vowel_cnt_map[left_char] = vowel_cnt_map.get(left_char, 0) - 1
                    if vowel_cnt_map[left_char] == 0:
                        vowel_cnt_map.pop(left_char)
                else:
                    consonant_cnt -= 1
                low += 1
            ans += low
        return ans
